overlord/ifacestate/apparmorprompting/common/common.go

```diff
@@ -0,0 +1,789 @@
+			return "", fmt.Errorf(`invalid duration: error parsing duration string: %s`, duration)
+		}
+		if parsedDuration <= 0 {
+			return "", fmt.Errorf(`invalid duration: duration must be greater than zero: %s`, duration)
+		}
+		expirationString = time.Now().Add(parsedDuration).Format(time.RFC3339)
+	default:
+		return "", fmt.Errorf(`invalid lifespan: "%v"`, lifespan)
+	}
+	return expirationString, nil
+}
+
+// Ensures that the given constraints, outcome, lifespan, and expiration are
+// valid for the given interface. If not, returns an error.
+func ValidateConstraintsOutcomeLifespanExpiration(iface string, constraints *Constraints, outcome OutcomeType, lifespan LifespanType, expiration string, currTime time.Time) error {
+	if err := constraints.ValidateForInterface(iface); err != nil {
+		return err
+	}
+	if err := ValidateOutcome(outcome); err != nil {
+		return err
+	}
+	return ValidateLifespanExpiration(lifespan, expiration, currTime)
+}
+
+// Ensures that the given constraints, outcome, lifespan, and duration are valid
+// for the given interface. If not, returns an error. Additionally, converts the
+// given duration to an expiration timestamp.
+func ValidateConstraintsOutcomeLifespanDuration(iface string, constraints *Constraints, outcome OutcomeType, lifespan LifespanType, duration string) (string, error) {
+	if err := constraints.ValidateForInterface(iface); err != nil {
+		return "", err
+	}
+	if err := ValidateOutcome(outcome); err != nil {
+		return "", err
+	}
+	return ValidateLifespanParseDuration(lifespan, duration)
+}
+
+// Returns true if the given pattern matches the given path. The pattern should
+// not contain groups, and should likely have been an output of ExpandPathPattern.
+//
+// Paths to directories are received with trailing slashes, but we don't want
+// to require the user to including a trailing '/' if they want to match
+// directories (and end their pattern with `{,/}` if they want to match both
+// directories and non-directories). Thus, we want to ensure that patterns
+// without trailing slashes match paths with trailing slashes. However,
+// patterns with trailing slashes should not match paths without trailing
+// slashes.
+//
+// The doublestar package has special cases for patterns ending in `/**` and
+// `/**/`: `/foo/**`, and `/foo/**/` both match `/foo` and `/foo/`. We want to
+// override this behavior to make `/foo/**/` not match `/foo`. We also want to
+// override doublestar to make `/foo` match `/foo/`.
+func PathPatternMatch(pattern string, path string) (bool, error) {
+	// Check the usual doublestar match first, in case the pattern is malformed
+	// and causes an error, and return the error if so.
+	matched, err := doublestar.Match(pattern, path)
+	if err != nil {
+		return false, err
+	}
+	// No matter if doublestar matched, return false if pattern ends in '/' but
+	// path is not a directory.
+	if len(pattern) > 0 && pattern[len(pattern)-1] == '/' && len(path) > 0 && path[len(path)-1] != '/' {
```

Comment:
Use `strings.HasSuffix()` to make sure trailing runes are `'/'`, rather than just trailing bytes.